marconi/src/Marconi/Index/Utxo.hs

```diff
@@ -239,227 +268,373 @@ instance FromField C.BlockNo where
-                    Shelley.ReferenceScript _
-                        (Shelley.ScriptInAnyLang
-                            (C.PlutusScriptLanguage (C.PlutusScriptV2) )
-                            script) -> (Just . C.serialiseToCBOR $ script, Just . C.hashScript $ script)    -- TODO this is hack to get pass build , Just . C.hashScript $ script)
-            in
-                Utxo {..}
+-- | Extract Utxos payload from Cardano Transaction
+-- Note, these Utxos will be decorated with additional data points to make a UtxoEvent
+--
+getUtxos :: (C.IsCardanoEra era) => Maybe TargetAddresses -> C.Tx era -> [Utxo]
+getUtxos maybeTargetAddresses (C.Tx txBody@(C.TxBody C.TxBodyContent {C.txOuts}) _) =
+  either (const []) addressDiscriminator (getUtxos' txOuts)
+  where
+    addressDiscriminator :: [Utxo] -> [Utxo]
+    addressDiscriminator = case maybeTargetAddresses of
+      Just targetAddresses -> filter (isAddressInTarget targetAddresses)
+      _                    -> id
+
+    getUtxos' :: C.IsCardanoEra era => [C.TxOut C.CtxTx era] -> Either C.EraCastError [Utxo]
+    getUtxos' = fmap (imap txoutToUtxo) . traverse (C.eraCast CurrentEra)
+
+    txoutToUtxo :: Int -> TxOut -> Utxo
+    txoutToUtxo ix out =
+      let
+        _txIx = C.TxIx $ fromIntegral ix
+        _txId = C.getTxId txBody
+        C.TxOut address' value' datum' refScript = out
+        _address = UtxoAddress . toAddr $ address'
+        _value = C.txOutValueToValue value'
+        _datumHash = case datum' of
+          (C.TxOutDatumHash _ d) -> Just d
+          _                      -> Nothing
+        _datum = case datum' of
+          (C.TxOutDatumInline _ d) -> Just d
+          _                        -> Nothing
+  -- tease out lang constraint as required by serialiseToCBOR
+        (_inlineScript, _inlineScriptHash) = case refScript of
+          Shelley.ReferenceScriptNone                             ->
+            (Nothing, Nothing)
+          Shelley.ReferenceScript _  (
+            Shelley.ScriptInAnyLang  (
+              C.SimpleScriptLanguage (C.SimpleScriptV1)) script)  ->
+             (Just . C.serialiseToCBOR $ script, Just . C.hashScript $ script)
+          Shelley.ReferenceScript _  (
+            Shelley.ScriptInAnyLang  (
+              C.SimpleScriptLanguage (C.SimpleScriptV2)) script)  ->
+            (Just . C.serialiseToCBOR $ script, Just . C.hashScript $ script)
+          Shelley.ReferenceScript _  (
+            Shelley.ScriptInAnyLang  (
+              C.PlutusScriptLanguage (C.PlutusScriptV1)) script)  ->
+            (Just . C.serialiseToCBOR $ script, Just . C.hashScript $ script)
+          Shelley.ReferenceScript _  (
+            Shelley.ScriptInAnyLang  (
+              C.PlutusScriptLanguage (C.PlutusScriptV2)) script)  ->
+            (Just . C.serialiseToCBOR $ script, Just . C.hashScript $ script)
+       in
+        Utxo {..}
+{-
+ TODO we need to remove spent from in-memory utxoEvents.
```

Comment:
We should also look into this scenario, but for now I don't see where this function is used. I don't think we should ever remove spent transactions from memory.